OCPNODE-2439: Add support for BYOPKI (#1658)
openshift/enhancements: enhancements/api-review/add-ClusterImagePolicy-and-ImagePolicy-for-Signature-Verification.md

	// https://github.com/sigstore/fulcio and https://github.com/sigstore/rekor
	// +optional
	FulcioCAWithRekor *FulcioCAWithRekor `json:"fulcioCAWithRekor,omitempty"`

    // PKI defines the root of trust based on Root CA(s) and corresponding intermediate certificates.
    // +optional

>     PKI *PKI `json:"pki,omitempty"`

Reviewer: nit: let's fix the indentation here.

Author: updated. thanks.

	// https://github.com/sigstore/fulcio and https://github.com/sigstore/rekor
	// +optional
	FulcioCAWithRekor *FulcioCAWithRekor `json:"fulcioCAWithRekor,omitempty"`

  // PKI defines the root of trust based on Root CA(s) and corresponding intermediate certificates.
  // +optional

>   PKI *PKI `json:"pki,omitempty"`

Reviewer 2: nit: seems like inconsistent indent whitespace vs. the previous `FulcioCAWithRekor` entry?

Author: Updated thanks.

Reviewer 3: (Still outstanding I think.)

	FulcioSubject PolicyFulcioSubject `json:"fulcioSubject,omitempty"`
}

// PKI defines the root of trust based on Root CA(s) and corresponding intermediate certificates.

> type PKI struct {

Reviewer 3: The `PKI` type is not used anywhere; it needs to be integrated into `PolicyRootOfTrust`.

Author: Oh sorry missed it, thanks for catching that.

	// CertificateAuthorityRootsData contains a certificate bundle PEM file containing one or more CA roots
	// in the PEM format.

> 	// caRootsData cannot be set if certificateChainData is set.

Reviewer 3: `certificateChainData` no longer exists (two instances, + validation annotations).

Author: thanks

> 	// +optional

Reviewer 3: Isn’t this required now?

Author: yes

	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateChainData == ''", message="CertificateAuthorityRoots cannot be set if CertificateChain is set"
	CertificateAuthorityRootsData []byte `json:"caRootsData,omitempty"`

	// CertificateAuthorityIntermediatesData contains a certificate bundle PEM file containing one or more intermediate certificates
	// in the PEM format.
	// caIntermediatesData cannot be set if certificateChainData is set.
	// caIntermediatesData requires CertificateAuthorityRoots to be set.
	// +optional
	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateChainData == ''", message="caIntermediatesData cannot be set if certificateChainData is set"

> 	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateAuthorityRootsData != ''", message="caIntermediatesData requires caRootsData to be set"

Reviewer 3: This validation also seems no longer applicable.

Author: agree

Reviewer 3: Now that `caRootsData` is mandatory, should this line be removed entirely?

Reviewer 3: … as well as the “requires CertificateAuthorityRoots to be set.” text above.

---

But then again, having the condition expressed here doesn’t really hurt anything.

	CertificateAuthorityIntermediatesData []byte `json:"caIntermediatesData,omitempty"`

	// CertificateIdentity must be set if any of the other fields are set.

> 	// +optional

Reviewer 3: Doesn’t this amount to the identity being mandatory?

---

But, also, see the comment on `CertificateIdentity`. What is the semantics we are proposing, _and then_ what are the k8s API evolution techniques we need to use?

Author: Maybe a rule like,

// +kubebuilder:validation:XValidation:rule="self.CertificateAuthorityRootsData == '' || self.PKICertificateSubject != nil", message="pkiCertificateSubject must be set if caRootsData is specified"

Reviewer 3: `caRootsData` is now required, so I think that implies to `PKICertificateSubject` being required as well.

Reviewer 3: … ah, https://github.com/openshift/enhancements/pull/1658#discussion_r1735227405 suggests a difference between the `+required` annotation and a `+kubebuilder:validation` annotation, WRT API rules compliance.

I’ll leave K8s experts to review that aspect — I’m primarily interested in whether we want the data, in the first version, to be mandatory or optional _in effect_.

Author: Marked it `required` for now, we can change it in future by moving the API to `v1beta2`

	// +kubebuilder:validation:XValidation:rule="(self.CertificateChainData == '' && self.CertificateAuthorityRootsData == '' && self.caIntermediatesData == '') || self != ''", message="certificateIdentity must be set if any of certificateChainData, caRootsData, or caIntermediatesData are set"
	// +kubebuilder:validation:XValidation:rule="self.Email != '' || self.Hostname != '' || self.IPAddress != ''", message="At least one of Email, Hostname, or IPAddress must be set in certificateIdentity"
	CertificateIdentity *CertificateIdentity `json:"certificateIdentity,omitempty"`

> }

Reviewer 3: Should we support Rekor timestamps (for short-lived certificates)?

Maybe that could be a separate future feature, depending on what the users need. See also the `CertificateIdentity` comment.

Author: We will consider Rekor for future, maybe when we go to `v1beta2`

// CertificateIdentity defines the identity type associated with the certificate identity

> type CertificateIdentity struct {

Reviewer 3: To be explicit, the proposal is that it will be possible to specify no matching requirements at all?

Compare https://issues.redhat.com/browse/RFE-5444?focusedId=24787146&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-24787146 , that’s only reasonable for a strictly single-purpose CA, and then … if we need a CA at all, we probably _have to_ consider revocation; that means the ability to require Rekor, and possible future extensibility to use non-Rekor timestamps as an alternative.

---

I didn’t look into K8s API evolution rules — can we start with making an identity mandatory, and relax that later?

Author: maybe we can keep it `optional` and enforce it using kubebuilder validation rules somehow?

Author: that way we don't have to change anything in actual API except the rules.

Author: // +kubebuilder:validation:XValidation:rule="self.CertificateAuthorityRootsData == '' || self.PKICertificateSubject != nil", message="pkiCertificateSubject must be set if caRootsData is specified"

Reviewer 3: Per a Slack discussion, this is an alpha API and we can update to `alpha2`, so maybe we don’t need to worry about the change between required/optional too much, _for now_.

Author: Marked it `required`, we can graduate the API to v1beta2 to make it `optional` if we wish. cc @wking

> type CertificateIdentity struct {

Reviewer 3: WRT the name of the object, it’s not actually dealing with the “certificate identity” (that’s something like (issuer + serial number)); this is dealing with the certificate _subject_.

Maybe something like `PKICertificateSubject`? And the field name in the containing object should be updated to match.

	// Email address associated with the certificate identity.
	// +optional
	Email string `json:"email,omitempty"`

	// Hostname associated with the certificate identity.
	// +optional
	Hostname string `json:"hostname,omitempty"`

	// IP address associated with the certificate identity.
	// +optional

> 	IPAddress string `json:"ipaddress,omitempty"`

Reviewer 3: I don’t know how IP address matching makes sense to administrators: “a certificate issued for 8.8.8.8 can sign our application”. It’s not _impossible_ but it seems all-around awkward to manage.

(… The Go certificate parser is already parsing IP addresses in SAN, so I guess adding this would not really introduce _risk_ and I wouldn’t block on this feature existing.)

Author: I agree, I will remove it.

	FulcioSubject PolicyFulcioSubject `json:"fulcioSubject,omitempty"`
}

// PKI defines the root of trust based on Root CA(s) and corresponding intermediate certificates.
type PKI struct {
	// CertificateAuthorityRootsData contains base64-encoded data of a certificate bundle PEM file, which contains one or more CA roots in the PEM format.
	// +required
	CertificateAuthorityRootsData []byte `json:"caRootsData,omitempty"`

	// CertificateAuthorityIntermediatesData base64-encoded data of a certificate bundle PEM file, which contains one or more intermediate certificates in the PEM format.
	// caIntermediatesData requires CertificateAuthorityRoots to be set.
	// +optional
	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateAuthorityRootsData != ''", message="caIntermediatesData requires caRootsData to be set"
	CertificateAuthorityIntermediatesData []byte `json:"caIntermediatesData,omitempty"`

> 	// PKICertificateSubject must be set if any of the other fields are set.

Reviewer: Let's add a note what this field does.

Author: Updated, thanksl

	// +required
  // +kubebuilder:validation:XValidation:rule="self.CertificateAuthorityRootsData == '' || self.PKICertificateSubject != nil", message="pkiCertificateSubject must be set if caRootsData is specified"

> 	PKICertificateSubject *PKICertificateSubject `json:"pkiCertificateSubject,omitempty"`

Reviewer: how about just using `CertificateSubject` since we're already in the `PKI` struct?
```suggestion
	CertificateSubject *CertificateSubject `json:"certificateSubject,omitempty"`
```

Reviewer 3: As far as correctness goes, sure, why not.

One concern to think about is that there are two audiences here:
- Users writing YAML: for them the `json` field names matter, and we can rely on the nesting structure to avoid repetition. So `certificateSubject` is better than `pkiCertificateSubject`
- Go programmers using the API package: all types are at the top level, so they should be unique standalone. So `PKICertificateSubject` is better than `CertificateSubject`, or maybe it should be `PKIImagePolicyCertificateSubject` in the extreme??
  - Note especially that the naming the top-level structure just `PKI` is not ideal… we can get away with that for `FulcioCAWithRekor`, but for `PKI` and the existing `PublicKey`… there’s some risk of clashes.

But, also, the YAML names are more important because they are user-facing; the Go type names are ~internal in a package with no stable API commitment (I think?) so if we ever encountered a conflict, we can rename things.

Reviewer 4: ```suggestion
	PKICertificateSubject *PKICertificateSubject `json:"certificateSubject,omitempty"`
```
Maybe only omit the PKI from the json annotation.

Reviewer 4: > Note especially that the naming the top-level structure just PKI is not ideal… we can get away with that for FulcioCAWithRekor, but for PKI and the existing PublicKey… there’s some risk of clashes.

will `CustomPKI` instead of `PKI` make a difference? we can explain in the comment this means a PKI not from Sigstore.

Reviewer 3: In the context of `openshift/api/config/v1alpha1` (or, worse, eventually, `…/v1`, I think `PKI` and `CustomPKI` are very similarly likely to clash — if there is going to be a clash, it’s going to be with some other non-`policy.json` PKI use.

See https://pkg.go.dev/github.com/openshift/api/config/v1 — there are things like `NamedCertificate` and `CertInfo`.

So I’d be tempted to use `ImagePolicy*` for all types (`ImagePolicyPKI`?), but that might well be an overreaction.

---

I do think the JSON/YAML field names are much more important; I’m not nearly as worried about the Go types.

	FulcioSubject PolicyFulcioSubject `json:"fulcioSubject,omitempty"`
}

// PKI defines the root of trust based on Root CA(s) and corresponding intermediate certificates.
type PKI struct {
	// CertificateAuthorityRootsData contains a certificate bundle PEM file containing one or more CA roots

> 	// in the PEM format.

Reviewer 4: I think `caRootsData` and  `caIntermediatesData`  are base64 encoded of the PEM format data? I will add a commit to address this.

Author: yes, thanks.

Reviewer 3: (Documenting `[]byte` fields is a bit awkward — it’s base64-encoded _in the JSON form_ but not in the Go values.

When other fields document the base64 aspect, doing that here makes sense to me.)

Author: Thanks.

	// +required
	CertificateAuthorityRootsData []byte `json:"caRootsData,omitempty"`

	// CertificateAuthorityIntermediatesData base64-encoded data of a certificate bundle PEM file, which contains one or more intermediate certificates
	// in the PEM format.
	// caIntermediatesData requires CertificateAuthorityRoots to be set.
	// +optional
	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateAuthorityRootsData != ''", message="caIntermediatesData requires caRootsData to be set"
	CertificateAuthorityIntermediatesData []byte `json:"caIntermediatesData,omitempty"`

	// PKICertificateSubject must be set if any of the other fields are set.
	// +optional
  // +kubebuilder:validation:XValidation:rule="self.CertificateAuthorityRootsData == '' || self.PKICertificateSubject != nil", message="pkiCertificateSubject must be set if caRootsData is specified"

> 	// +kubebuilder:validation:XValidation:rule="self.Email != '' || self.Hostname != ''", message="At least one of Email or Hostname must be set in pkiCertificateSubject"

Reviewer 3: (Is it idiomatic to have the validation here? I’d expect it to exist inside the `PKICertificateSubject` type… but I also didn’t check that it is possible.)

Author: makes sense, I moved it inside `PKICertificateSubject`

	PKICertificateSubject *PKICertificateSubject `json:"pkiCertificateSubject,omitempty"`
}

> // PKICertificateSubject defines the identity type associated with the certificate identity

Reviewer 3: ```suggestion
// PKICertificateSubject defines the requirements imposed on the subject to which the certificate was issued.
```

---

The fields below should also talk about subject, not identity — and perhaps explicitly point say that this  matches against the corresponding “subject alternative name” fields.

Author: Updated, thanks.

Reviewer 3: Still s/certificate identity/certificate subject/ in the fields of `…CertficateSubject`, please.

Reviewer 4: I just saw this review, I can address and let people review the API comments/validation/naming details in the openshift/api PR later. cc @harche
